global.CSS might not exist in the browser and def. doesn't in node

kof opened the pull request.

HenriBeck commented on packages/jss/src/utils/escape.js:
```diff
@@ -2,14 +2,12 @@ const env = process.env.NODE_ENV
 
 const escapeRegex = /([[\].#*$><+~=|^:(),"'`])/g
 
+const nativeEscape = typeof CSS === 'undefined' ? CSS.escape : undefined
```
Should be `typeof CSS !== 'undefined'`

HenriBeck replied:
Also could be `const escape = typeof CSS !== 'undefined' ? CSS.escape : (str) => str.replace(escapeRegex, '\\$1')`

That way we would just call `escape(str)` then.

kof replied:
already fixed the first one, we need an additional wrapper around this escape function to avoid escaping in production

kof replied:
actually I need to shift as much as possible inside of that wrapping function, because it will be removed from the bundle

HenriBeck replied:
yeah, but instead of doing the check if nativeEscape exists we would just call a function then

HenriBeck replied:
Why will it be removed from the bundle?

kof replied:
because of the if production check

HenriBeck replied:
```js
const escape = typeof CSS !== 'undefined' ? CSS.escape : str => str.replace(escapeRegex, '\\$1')

export default str => {
  // We don't need to escape it in production, because we are not using user's
  // input for selectors, we are generating a valid selector.
  if (env === 'production') return str
  
  return escape(str)
}
```

HenriBeck commented on packages/jss/src/utils/escape.js:
```diff
@@ -2,14 +2,12 @@ const env = process.env.NODE_ENV
 export default str => {
   // We don't need to escape it in production, because we are not using user's
   // input for selectors, we are generating a valid selector.
   if (env === 'production') return str
```
Quick question about that: What happens if the user has a custom generate-id and passed some non-valid css selector.

kof replied:
in that case its up to the user, because we want to have max perf, escaping has a cost

HenriBeck replied:
We should maybe note that in the docs about a custom generate-id if it's not already there.

kof replied:
yeah, the resulting id should be a selector conform string

kof commented:
updated again, @HenriBeck check . out min bundle, the entire fn body is gone

HenriBeck commented:
Do we actually need to inline all of the code to get it removed? The function should be in production `str => str`. Then the variables outside aren't used anymore so they should also be stripped away correct?

kof commented:
not sure, will check

kof commented:
if its on top: `"undefined"!=typeof CSS&&CSS.escape;var k=function(e){return e}`

HenriBeck commented:
weird, alright.

kof commented:
it probably has to stay because minifier doesn't know if property access can be a getter, so that code might actually do something

kof commented:
I don't get it, CI fails on bundle size checks, I cloned the repo fresh and everything passes locally, I am gonna try merging it and hope its some weird CI problem that won't be there on master

kof commented:
nope, still fails, I need it on master though for @oliviertassinari , lets figure why it fails on master retrospectively

oliviertassinari commented on packages/jss/src/utils/escape.js:
```diff
@@ -1,15 +1,10 @@
-const escapeRegex = /([[\].#*$><+~=|^:(),"'`])/g
-
 export default str => {
   // We don't need to escape it in production, because we are not using user's
```
On Material-UI, we still escape in production as people can use a global CSS approach, where they provide a name to each style sheet.

oliviertassinari replied:
It's what we use at https://www.onepixel.com/en/ to get some decent caching performance.

kof replied:
If they do this, they also use valid strings for class names, why do you still need escaping there?

kof replied:
yes, react-jss does too, exactly the same story

oliviertassinari commented on packages/jss/src/utils/escape.js:
```diff
@@ -1,15 +1,10 @@
-  if (CSS && CSS.escape) {
-    return CSS.escape(str)
-  }
+  const escapeRegex = /([[\].#*$><+~=|^:(),"'`])/g
```
Should the regexp be moved outside of the function to provide better pers in dev?

oliviertassinari replied:
Also, we uses `/([[\].#*$><+~=|^:(),"'`\s])/g;` on Material-UI side to solve some other edge cases.

kof replied:
I usually do that, but actually engine already does optimizations there, its not really necessarily and it won't be removed from the bundle afaik, if it does get removed we can move it to the top

kof replied:
will update the regex with yours, it fixes the issue with spaces we already got reported and backticks

kof replied:
oh just realized you don't have the backtick in the regex, you should add it too

kof replied:
just looked again yes, I already added it too!

kof replied:
our regexes are unified now, does mui has tests for all cases that regex covers? If yes I would copy them to make sure we don't break them by accident

HenriBeck commented:
I will check it this evening or tomorrow

oliviertassinari commented on packages/jss-preset-default/.size-snapshot.json:
```diff
@@ -1,13 +1,13 @@
-    "bundled": 67203,
-    "minified": 23226,
-    "gzipped": 6969
+    "bundled": 30500,
```
I would encourage the usage if [size-limit](https://github.com/ai/size-limit) over having to commit the new values every time a pull request is done. it's wasteful.

kof replied:
size limit has also a fixed limit, which you have to fix if you increase, if you make it too flexible - than you don't control the size. I think the point of it is to make you aware of every size change on   for every change

kof replied:
also we have different numbers here, which might behave differently, its interesting to see how minified/gziped size changes compared to raw, those numbers are not linear

kof replied:
well, they only need to run yarn build, thats it, if not, we can as well do it, its not too bad so far, but if it gets annoying we can add a git hook cc @TrySound

kof commented:
fixed on master, also moved the regex to the top and added whitespace handling, regex is getting removed in production, since it makes no side effects

oliviertassinari commented:
🎉
